app/uk/gov/hmrc/gform/builder/BuilderController.scala

```diff
@@ -380,15 +382,91 @@ class BuilderController(
     }
 
   def updateFormComponent(formTemplateRawId: FormTemplateRawId, formComponentId: FormComponentId) =
     updateAction(formTemplateRawId) { (formTemplateRaw, requestBody) =>
-      BuilderSupport
-        .updateFormComponent(formTemplateRaw, requestBody, formComponentId)
-        .map(formTemplateRaw => (formTemplateRaw, Results.Ok(formTemplateRaw.value)))
+      import Decoders._
+      for {
+        componentUpdateRequest <-
+          requestBody.as[ComponentUpdateRequest].leftMap(e => BuilderError.CirceDecodingError(e))
+        sectionUpdatedFormTemplateRaw <-
+          componentUpdateRequest.sectionDetails
+            .map(s => BuilderSupport.updateSectionByPath(formTemplateRaw, s.sectionPath, s.sectionData))
+            .getOrElse(Right(formTemplateRaw))
+        componentUpdatedFormTemplateRaw <-
+          BuilderSupport
+            .updateFormComponent(sectionUpdatedFormTemplateRaw, componentUpdateRequest.payload, formComponentId)
+      } yield (componentUpdatedFormTemplateRaw, Results.Ok(componentUpdatedFormTemplateRaw.value))
     }
 }
+
+case class SectionDetails(
+  sectionData: Json,
+  sectionPath: String
+)
+case class ComponentUpdateRequest(
+  payload: Json,
+  sectionDetails: Option[SectionDetails]
+)
+
+case class ComponentDetails(
+  componentData: Json
+)
+case class SectionUpdateRequest(
+  payload: Json,
+  componentDetails: Option[ComponentDetails]
+)
+
+object Decoders {
+  implicit val sectionDetailsDecoder: Decoder[SectionDetails] = new Decoder[SectionDetails] {
+    final def apply(cursor: HCursor): Decoder.Result[SectionDetails] =
+      for {
+        sectionData <- cursor.get[Json]("sectionData").map(_.dropNullValues)
+        sectionPath <- cursor.get[String]("sectionPath")
+      } yield SectionDetails(sectionData, sectionPath)
+  }
+
+  implicit val componentUpdateDecoder: Decoder[ComponentUpdateRequest] = new Decoder[ComponentUpdateRequest] {
+    final def apply(c: HCursor): Decoder.Result[ComponentUpdateRequest] = {
+      val payload = c.downField("sectionDetails").delete.top.getOrElse(c.value)
+      val sectionDetails = c.get[Option[SectionDetails]]("sectionDetails")
+      sectionDetails.map(s => ComponentUpdateRequest(payload, s))
+    }
+  }
+
+  implicit val componentDetailsDecoder: Decoder[ComponentDetails] = new Decoder[ComponentDetails] {
+    final def apply(cursor: HCursor): Decoder.Result[ComponentDetails] =
+      cursor.get[Json]("componentData").map(p => ComponentDetails(p.dropNullValues))
+  }
+
+  implicit val sectionUpdateDecoder: Decoder[SectionUpdateRequest] = new Decoder[SectionUpdateRequest] {
+    final def apply(c: HCursor): Decoder.Result[SectionUpdateRequest] = {
+      val payload = c.downField("componentDetails").delete.top.getOrElse(c.value)
+      val componentDetails = c.get[Option[ComponentDetails]]("componentDetails")
+      componentDetails.map(s => SectionUpdateRequest(payload, s))
+    }
```

Comment:
Would it be possible to derive these decoders?

Reply:
We can make derive SectionDetails and ComponentDetails but not SectionUpdateRequest and ComponentUpdateRequest like here:
```
case class SectionDetails(sectionData: Json, sectionPath: String)

object SectionDetails {
  implicit val decodeSectionDetails: Decoder[SectionDetails] = deriveDecoder[SectionDetails]
}

case class ComponentDetails(componentData: Json)

object ComponentDetails {
  implicit val decodeComponentDetails: Decoder[ComponentDetails] = deriveDecoder[ComponentDetails]
}

case class ComponentUpdateRequest(payload: Json, sectionDetails: Option[SectionDetails])

case class SectionUpdateRequest(payload: Json, componentDetails: Option[ComponentDetails])

object Decoders {
  implicit val componentUpdateDecoder: Decoder[ComponentUpdateRequest] = new Decoder[ComponentUpdateRequest] {
    final def apply(c: HCursor): Decoder.Result[ComponentUpdateRequest] = {
      val payload = c.downField("sectionDetails").delete.top.getOrElse(c.value)
      val sectionDetails = c.get[Option[SectionDetails]]("sectionDetails")
      sectionDetails.map(s => ComponentUpdateRequest(payload, s))
    }
  }
  implicit val sectionUpdateDecoder: Decoder[SectionUpdateRequest] = new Decoder[SectionUpdateRequest] {
    final def apply(c: HCursor): Decoder.Result[SectionUpdateRequest] = {
      val payload = c.downField("componentDetails").delete.top.getOrElse(c.value)
      val componentDetails = c.get[Option[ComponentDetails]]("componentDetails")
      componentDetails.map(s => SectionUpdateRequest(payload, s))
    }
  }
}
```

Reply:
However we can fully derive it if we completely change the current update requests payload to have optional fields for both section and component payloads

Reply:
Why both of them need to be optional?

Reply:
One request is for section update an I modified a request to have optionally update component.
Anohter request is for component update and I modified it to optionally update section.
If we are about to merge then then both should be optional I think